File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofServices/AgentService.swift
import Foundation
import Supabase
import WagerproofModels

/// Port of `wagerproof-mobile/services/agentService.ts`.
///
/// All reads/writes hit the **main** Supabase project (`gnjrklxotmbvnxbnnqgq`)
/// — the CFB project never touches `avatar_*` tables. RLS policies on
/// `avatar_profiles` constrain owner reads to `user_id = auth.uid()` plus
/// anonymous reads when `is_public = true`.
///
/// We deliberately split the column projection mirror so list calls fetch the
/// "slim" set (no JSONB) and detail calls fetch `*`. Matches the RN service
/// byte-for-byte — see `AGENT_LIST_COLUMNS` / `AGENT_DETAIL_COLUMNS`.
public enum AgentService {
    /// List projection — excludes `personality_params` + `custom_insights` so
    /// the grid query stays small. Same as `AGENT_LIST_COLUMNS` in RN.
    private static let listColumns = "id, user_id, name, avatar_emoji, avatar_color, preferred_sports, archetype, is_public, is_active, created_at, updated_at, auto_generate, auto_generate_time, auto_generate_timezone, is_widget_favorite, last_generated_at, last_auto_generated_at, owner_last_active_at, daily_generation_count, last_generation_date"

    /// Performance cache projection. Mirrors `PERF_COLUMNS` in RN.
    private static let performanceColumns = "avatar_id, wins, losses, pushes, total_picks, win_rate, net_units, current_streak, best_streak, worst_streak, last_calculated_at"

    /// Fetch a user's agents joined with their cached performance rows.
    /// Mirrors `fetchUserAgents` in `agentService.ts`.
    public static func fetchUserAgents(userId: String) async throws -> [AgentWithPerformance] {
        let main = await MainSupabase.shared.client

        // 1) Agents — slim list projection so we don't pull JSONB.
        let agents: [Agent] = try await main
            .from("avatar_profiles")
            .select(listColumns)
            .eq("user_id", value: userId)
            .order("created_at", ascending: false)
            .execute()
            .value

        guard !agents.isEmpty else { return [] }

        // 2) Performance cache rows for every fetched agent. Failure here
        // doesn't blank the list — we return agents with `performance: nil`
        // (matches RN's defensive try/catch).
        var perfMap: [String: AgentPerformance] = [:]
        let ids = agents.map { $0.id }
        do {
            let perfs: [AgentPerformance] = try await main
                .from("avatar_performance_cache")
                .select(performanceColumns)
                .in("avatar_id", values: ids)
                .execute()
                .value
            for p in perfs { perfMap[p.avatarId] = p }
        } catch {
            // swallow — agents still render.
        }

        return agents.map { AgentWithPerformance(agent: $0, performance: perfMap[$0.id]) }
    }

    /// Fetch a single agent (with all JSONB fields) and its performance row.
    /// Mirrors `fetchAgentById`.
    public static func fetchAgent(id: String) async throws -> AgentWithPerformance? {
        let main = await MainSupabase.shared.client
        let agents: [Agent] = try await main
            .from("avatar_profiles")
            .select()
            .eq("id", value: id)
            .limit(1)
            .execute()
            .value
        guard let agent = agents.first else { return nil }

        let perf: AgentPerformance? = await {
            do {
                let rows: [AgentPerformance] = try await main
                    .from("avatar_performance_cache")
                    .select(performanceColumns)
                    .eq("avatar_id", value: id)
                    .limit(1)
                    .execute()
                    .value
                return rows.first
            } catch {
                return nil
            }
        }()

        return AgentWithPerformance(agent: agent, performance: perf)
    }

    /// Delete an agent. Cascades pick + performance rows via FK ON DELETE
    /// CASCADE. Mirrors `deleteAgent`.
    public static func delete(agentId: String) async throws {
        let main = await MainSupabase.shared.client
        _ = try await main
            .from("avatar_profiles")
            .delete()
            .eq("id", value: agentId)
            .execute()
    }

    /// Toggle the `is_active` flag (used to pause autopilot without deleting).
    public static func setActive(agentId: String, isActive: Bool) async throws {
        let main = await MainSupabase.shared.client
        struct Patch: Encodable { let is_active: Bool; let updated_at: String }
        let patch = Patch(is_active: isActive, updated_at: Self.nowISO())
        _ = try await main
            .from("avatar_profiles")
            .update(patch)
            .eq("id", value: agentId)
            .execute()
    }

    /// Toggle `is_public`. Lands properly in B15 (detail/settings) — we expose
    /// it now so the long-press menu on AgentIdCard can flip it.
    public static func setPublic(agentId: String, isPublic: Bool) async throws {
        let main = await MainSupabase.shared.client
        struct Patch: Encodable { let is_public: Bool; let updated_at: String }
        let patch = Patch(is_public: isPublic, updated_at: Self.nowISO())
        _ = try await main
            .from("avatar_profiles")
            .update(patch)
            .eq("id", value: agentId)
            .execute()
    }

    /// Toggle `auto_generate` (autopilot on/off).
    public static func setAutoGenerate(agentId: String, autoGenerate: Bool) async throws {
        let main = await MainSupabase.shared.client
        struct Patch: Encodable { let auto_generate: Bool; let updated_at: String }
        let patch = Patch(auto_generate: autoGenerate, updated_at: Self.nowISO())
        _ = try await main
            .from("avatar_profiles")
            .update(patch)
            .eq("id", value: agentId)
            .execute()
    }

    /// Create a new agent via the `agent-authorized-action-v1` edge function.
    /// Mirrors `createAgent` in `agentService.ts:174` → `invokeAgentAuthorizedAction`
    /// with `action: 'create_agent'`. The edge function performs RevenueCat
    /// gating + Zod validation server-side; if the user can't create another
    /// agent (free tier saturated, etc.) it returns a 4xx that surfaces as a
    /// thrown error here.
    public static func create(input: CreateAgentInput) async throws -> Agent {
        let payload = CreateAgentEdgePayload(
            action: "create_agent",
            data: input
        )
        return try await invokeAgentAuthorizedAction(payload: payload, fallback: "Failed to create agent")
    }

    /// Wraps the supabase edge invoke with the explicit bearer header behavior
    /// from `agentAuthorizedActions.ts:60-77`. Without the explicit header the
    /// SDK occasionally drops auth on `verify_jwt=false` functions and returns
    /// 401 even with an active session.
    private static func invokeAgentAuthorizedAction<Payload: Encodable, Response: Decodable>(
        payload: Payload,
        fallback: String
    ) async throws -> Response {
        let main = await MainSupabase.shared.client
        let session = try await main.auth.session
        // Explicit Authorization mirrors the RN explicit-bearer workaround for
        // SDK regressions where the auto-attached header gets dropped on
        // `verify_jwt=false` functions. See agentAuthorizedActions.ts:60-77.
        let options = FunctionInvokeOptions(
            headers: ["Authorization": "Bearer \(session.accessToken)"],
            body: payload
        )

        let envelope: AgentAuthorizedActionResponse<Response> = try await main.functions.invoke(
            "agent-authorized-action-v1",
            options: options
        )
        if let data = envelope.data, envelope.success == true {
            return data
        }
        throw NSError(
            domain: "AgentAuthorizedAction",
            code: -1,
            userInfo: [NSLocalizedDescriptionKey: envelope.error ?? fallback]
        )
    }

    private static func nowISO() -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f.string(from: Date())
    }
}

// MARK: - Edge function payloads

/// Encoded payload for the `create_agent` action. Matches the RN type union in
/// `services/agentAuthorizedActions.ts:3-30` byte-for-byte — `action` discriminator
/// + nested `data` object.
private struct CreateAgentEdgePayload<T: Encodable>: Encodable {
    let action: String
    let data: T
}

/// Standard envelope returned by `agent-authorized-action-v1` — matches the
/// `AgentAuthorizedActionSuccess<T> | Failure` union from the RN client.
private struct AgentAuthorizedActionResponse<T: Decodable>: Decodable {
    let success: Bool?
    let data: T?
    let error: String?
}

/// Input payload for creating an agent — the wizard collects these fields and
/// the edge function validates server-side via the Zod schema mirrored at
/// `wagerproof-mobile/types/agent.ts::CreateAgentSchema`. Field-for-field
/// match so the same JSON shape the RN app posts is what we post here.
public struct CreateAgentInput: Encodable, Sendable {
    public let name: String
    public let avatarEmoji: String
    public let avatarColor: String
    public let preferredSports: [AgentSport]
    public let archetype: AgentArchetype?
    public let personalityParams: AgentPersonalityParams
    public let customInsights: AgentCustomInsights
    public let autoGenerate: Bool
    public let autoGenerateTime: String
    public let autoGenerateTimezone: String

    public init(
        name: String,
        avatarEmoji: String,
        avatarColor: String,
        preferredSports: [AgentSport],
        archetype: AgentArchetype?,
        personalityParams: AgentPersonalityParams,
        customInsights: AgentCustomInsights,
        autoGenerate: Bool,
        autoGenerateTime: String,
        autoGenerateTimezone: String
    ) {
        self.name = name
        self.avatarEmoji = avatarEmoji
        self.avatarColor = avatarColor
        self.preferredSports = preferredSports
        self.archetype = archetype
        self.personalityParams = personalityParams
        self.customInsights = customInsights
        self.autoGenerate = autoGenerate
        self.autoGenerateTime = autoGenerateTime
        self.autoGenerateTimezone = autoGenerateTimezone
    }

    enum CodingKeys: String, CodingKey {
        case name
        case avatarEmoji = "avatar_emoji"
        case avatarColor = "avatar_color"
        case preferredSports = "preferred_sports"
        case archetype
        case personalityParams = "personality_params"
        case customInsights = "custom_insights"
        case autoGenerate = "auto_generate"
        case autoGenerateTime = "auto_generate_time"
        case autoGenerateTimezone = "auto_generate_timezone"
    }
}

File: wagerproof-android-native/core/services/src/main/kotlin/com/wagerproof/core/services/AgentService.kt
package com.wagerproof.core.services

import com.wagerproof.core.models.Agent
import com.wagerproof.core.models.AgentArchetype
import com.wagerproof.core.models.AgentArchetypeSerializer
import com.wagerproof.core.models.AgentCustomInsights
import com.wagerproof.core.models.AgentPerformance
import com.wagerproof.core.models.AgentPersonalityParams
import com.wagerproof.core.models.AgentSport
import com.wagerproof.core.models.AgentWithPerformance
import com.wagerproof.core.models.serialization.WagerproofJson
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.jsonObject

/**
 * CRUD on `avatar_profiles` (Main project). Port of iOS `AgentService.swift` /
 * RN `agentService.ts`. RLS constrains owner reads to `user_id = auth.uid()`
 * plus anonymous reads when `is_public = true`.
 */
object AgentService {

    // List projection excludes personality_params + custom_insights JSONB so
    // the grid query stays small. Byte-for-byte with RN AGENT_LIST_COLUMNS.
    private const val LIST_COLUMNS =
        "id, user_id, name, avatar_emoji, avatar_color, preferred_sports, archetype, " +
            "is_public, is_active, created_at, updated_at, auto_generate, auto_generate_time, " +
            "auto_generate_timezone, is_widget_favorite, last_generated_at, last_auto_generated_at, " +
            "owner_last_active_at, daily_generation_count, last_generation_date"

    private const val PERFORMANCE_COLUMNS =
        "avatar_id, wins, losses, pushes, total_picks, win_rate, net_units, " +
            "current_streak, best_streak, worst_streak, last_calculated_at"

    /** A user's agents joined with their cached performance rows. */
    suspend fun fetchUserAgents(userId: String): List<AgentWithPerformance> {
        val agents = SupabaseClients.main.from("avatar_profiles")
            .select(columns = Columns.raw(LIST_COLUMNS)) {
                filter { eq("user_id", userId) }
                order("created_at", Order.DESCENDING)
            }
            .decodeList<Agent>()
        if (agents.isEmpty()) return emptyList()

        // Perf-cache failure never blanks the list — agents render with
        // performance = null (matches the RN/iOS defensive try/catch).
        val perfMap = runCatching {
            SupabaseClients.main.from("avatar_performance_cache")
                .select(columns = Columns.raw(PERFORMANCE_COLUMNS)) {
                    filter { isIn("avatar_id", agents.map { it.id }) }
                }
                .decodeList<AgentPerformance>()
        }.getOrDefault(emptyList()).associateBy { it.avatarId }

        return agents.map { AgentWithPerformance(agent = it, performance = perfMap[it.id]) }
    }

    /** Single agent with all JSONB fields plus its performance row (null on perf error). */
    suspend fun fetchAgent(id: String): AgentWithPerformance? {
        val agent = SupabaseClients.main.from("avatar_profiles")
            .select {
                filter { eq("id", id) }
                limit(1)
            }
            .decodeList<Agent>()
            .firstOrNull() ?: return null

        val perf = runCatching {
            SupabaseClients.main.from("avatar_performance_cache")
                .select(columns = Columns.raw(PERFORMANCE_COLUMNS)) {
                    filter { eq("avatar_id", id) }
                    limit(1)
                }
                .decodeList<AgentPerformance>()
                .firstOrNull()
        }.getOrNull()

        return AgentWithPerformance(agent = agent, performance = perf)
    }

    /** Delete an agent; FK ON DELETE CASCADE removes picks + performance rows. */
    suspend fun delete(agentId: String) {
        SupabaseClients.main.from("avatar_profiles").delete {
            filter { eq("id", agentId) }
        }
    }

    /** Toggle `is_active` (pause autopilot without deleting). */
    suspend fun setActive(agentId: String, isActive: Boolean) =
        patchFlag(agentId, "is_active", isActive)

    /** Toggle `is_public`. */
    suspend fun setPublic(agentId: String, isPublic: Boolean) =
        patchFlag(agentId, "is_public", isPublic)

    /** Toggle `auto_generate` (autopilot on/off). */
    suspend fun setAutoGenerate(agentId: String, autoGenerate: Boolean) =
        patchFlag(agentId, "auto_generate", autoGenerate)

    private suspend fun patchFlag(agentId: String, column: String, value: Boolean) {
        SupabaseClients.main.from("avatar_profiles").update({
            set(column, value)
            set("updated_at", nowISO())
        }) {
            filter { eq("id", agentId) }
        }
    }

    /**
     * Create a new agent via the `create_agent` edge action. The edge function
     * performs RevenueCat gating + Zod validation server-side; a rejection
     * surfaces as [AgentAuthorizedActionsService.ActionException.Server].
     */
    suspend fun create(input: CreateAgentInput): Agent {
        val data = WagerproofJson
            .encodeToJsonElement(CreateAgentInput.serializer(), input)
            .jsonObject
        return AgentAuthorizedActionsService.createAgent(data)
    }

    // Swift ISO8601DateFormatter(.withInternetDateTime, .withFractionalSeconds):
    // always UTC with fractional seconds, even when millis are zero.
    private val isoFormatter =
        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).withZone(ZoneOffset.UTC)

    private fun nowISO(): String = isoFormatter.format(Instant.now())
}

/**
 * Input payload for `create_agent`. Field-for-field match with the Zod schema
 * mirrored at `wagerproof-mobile/types/agent.ts::CreateAgentSchema` (snake_case
 * keys) so all clients post the same JSON shape. Not in core/models because
 * only the create flow uses it (mirrors its Swift home in AgentService.swift).
 */
@Serializable
data class CreateAgentInput(
    val name: String,
    @SerialName("avatar_emoji") val avatarEmoji: String,
    @SerialName("avatar_color") val avatarColor: String,
    @SerialName("preferred_sports") val preferredSports: List<AgentSport>,
    @Serializable(with = AgentArchetypeSerializer::class)
    val archetype: AgentArchetype? = null,
    @SerialName("personality_params") val personalityParams: AgentPersonalityParams,
    @SerialName("custom_insights") val customInsights: AgentCustomInsights,
    @SerialName("auto_generate") val autoGenerate: Boolean,
    @SerialName("auto_generate_time") val autoGenerateTime: String,
    @SerialName("auto_generate_timezone") val autoGenerateTimezone: String,
)
